File: algorithms/1_easy/C/04030_check_ascii_palindromic/solution.c
typedef ptrdiff_t isize;

inline static unsigned char rev_bits(unsigned char ch) {
    unsigned char rev_ch = 0;

    for (int i = 0; i < 8; i += 1) {
        rev_ch <<= 1;
        rev_ch |= (ch & 1);
        ch >>= 1;
    }

    return rev_ch;
}

bool isPalindromic(const char *restrict s) {
    if (s == NULL || s[0] == '\0') { return true; }

    const isize len = (isize)strlen(s);
    isize hi = len - 1;
    isize lo = 0;

    while (lo <= hi) {
        if (rev_bits((unsigned char)s[lo]) != (unsigned char)s[hi]) {
            return false;
        }
        lo += 1;
        hi -= 1;
    }

    return true;
}

File: algorithms/1_easy/C/04030_check_ascii_palindromic/solution.rs
impl Solution {
    #[inline(always)]
    fn rev_bits(mut byte: u8) -> u8 {
        let mut rev_byte = 0_u8;

        for _ in 0..u8::BITS {
            rev_byte <<= 1;
            rev_byte |= (byte & 1);
            byte >>= 1;
        }

        rev_byte
    }

    pub fn is_palindromic(s: String) -> bool {
        let bytes = s.as_bytes();
        let len = bytes.len();
        let mut hi = (len as isize) - 1;
        let mut lo = 0_isize;

        while lo <= hi {
            if Self::rev_bits(bytes[lo as usize]) != bytes[hi as usize] {
                return false;
            }
            lo += 1;
            hi -= 1;
        }

        true
    }
}
